### PR TITLE
Support atomic testing and resource diff based oracle

### PR DESCRIPTION
**Pull request by laphets**
What's included,
- atomic bug test facility
- rabbitmq and casskop atomic bug test
- csi driver for pvc resizing on one node cluster
- resource diff based oracle (currently do not run twice, but use predefined rules for ignore)

### CODE REVIEW

File: .github/workflows/docker-publish.yml [hunk 1]
```diff
@@ -93,6 +97,14 @@ jobs:
       - name: Build Image
         run: |
           python3 build.py -p cass-operator -m time-travel -d $IMAGE_NAMESPACE
+      - name: Sieve Learn
+        run: |
+          python3 sieve.py -p cass-operator -t recreate -m learn -d $IMAGE_NAMESPACE
```
Comment (marshtompsxd) [inline .github/workflows/docker-publish.yml:102]: Can you do `-s learn` here?
The most recent version only accepts `-s learn` if you want to run the learning stage

File: .github/workflows/docker-publish.yml [hunk 2]
```diff
@@ -138,14 +149,24 @@ jobs:
       - name: Build Image
         run: |
           python3 build.py -p xtradb-operator -m time-travel -d $IMAGE_NAMESPACE
+      - name: Sieve Learn
+        run: |
+          python3 sieve.py -p xtradb-operator -t recreate -m learn -d $IMAGE_NAMESPACE
```
Comment (marshtompsxd) [inline .github/workflows/docker-publish.yml:154]: same as above. Please use `-s learn` for learning stage

File: .github/workflows/docker-publish.yml [hunk 3]
```diff
@@ -171,15 +192,22 @@ jobs:
...
       - name: Install Kind
         run: |
           go get sigs.k8s.io/kind
           kind
       - name: Build Image
         run: |
           python3 build.py -p mongodb-operator -m time-travel  -d $IMAGE_NAMESPACE
+      - name: Sieve Learn
+        run: |
+          python3 sieve.py -p mongodb-operator -t recreate -m learn -d $IMAGE_NAMESPACE
```
Comment (marshtompsxd) [inline .github/workflows/docker-publish.yml:205]: same above

File: analyze.py [hunk 4]
```diff
@@ -427,6 +427,39 @@ def generate_obs_gap_yaml(triggering_points, path, project, node_ignore):
             os.path.join(path, "obs-gap-config-%s.yaml" % (str(i))), "w"), sort_keys=False)
     print("Generated %d obs-gap config(s) in %s" % (i, path))
 
+def generate_atomic_yaml(triggering_points, path, project, node_ignore):
+    yaml_map = {}
+    yaml_map["project"] = project
+    yaml_map["stage"] = "test"
+    yaml_map["mode"] = "atomic"
+    yaml_map["straggler"] = controllers.straggler
```
[the remark this answers is not in the document]
Comment (marshtompsxd) [reply analyze.py:435]: Same for front runner

File: analyze.py [hunk 5]
```diff
@@ -469,6 +502,9 @@ def analyze_trace(project, log_dir, analysis_mode, generate_oracle=True, generat
     if not generate_config:
         two_sided = False
         use_sql = False
+    # Disable sql feature for obs gap / atomic mode for now
+    if analysis_mode == "obs-gap" or analysis_mode == "atomic":
```
Comment (marshtompsxd) [inline analyze.py:506]: I have turned down sql by default, so you can delete this check

File: csi-driver/install.sh [hunk 6]
```diff
@@ -0,0 +1,37 @@
+#!/bin/sh
```
Comment (marshtompsxd) [inline csi-driver/install.sh:1]: Are the files in `csi-driver/deploy` cloned from some repo or written by you?

File: oracle.py [hunk 7]
```diff
@@ -208,4 +296,55 @@ def check(learned_side_effect, learned_status, testing_side_effect, testing_stat
                 testing_config)
         bug_report = "\n[BUG REPORT]\n" + bug_report
         print(bug_report)
+
+    if learned_resources != None:
+        look_for_resouces_diff(learned_resources, testing_resources)
```
Comment (marshtompsxd) [inline oracle.py:301]: I guess we should generate bug report according to `look_for_resources_diff`, right?

Comment (marshtompsxd) [reply oracle.py:301]: and also generate alarm

File: sieve.py [hunk 8]
```diff
@@ -119,6 +119,10 @@ def setup_cluster(project, stage, mode, test_config, docker_repo, docker_tag, nu
         os.system("docker pull %s" % (image))
         os.system(kind_load_cmd)
 
+    if num_workers == 0:
```
Comment (marshtompsxd) [inline sieve.py:122]: Let's make the csi-driver an option in the test suite and enable it for the rabbitmq pvc resize only, instead of checking the number of workers

File: sonar-server/atomic-server.go [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+		checkError(err)
+		if len(pods.Items) != 0 {
+			log.Printf("operator pod not deleted yet\n")
+		} else {
+			log.Printf("operator pod gets deleted\n")
+			break
+		}
+	}
+
+	time.Sleep(time.Second * 15)
```
Comment (marshtompsxd) [inline sonar-server/atomic-server.go:175]: As discussed yesterday, let's start with the most simple form without the 15s hardcoded sleep.

File: sonar-server/obs-gap-server.go [hunk 10]
```diff
@@ -122,8 +122,8 @@ func (s *obsGapServer) getEventResourceNamespace(event map[string]interface{}) s
 	}
 }
 
-func (s *obsGapServer) isSameTarget(currentEvent map[string]interface{}) bool {
-	return s.getEventResourceName(currentEvent) == s.ceName && s.getEventResourceNamespace(currentEvent) == s.ceNamespace
+func isSameTarget(e1, e2 map[string]interface{}) bool {
```
Comment (marshtompsxd) [inline sonar-server/obs-gap-server.go:125]: Please revert this change. `getEventResourceName(e1) == getEventResourceName(e2)` does not always work because the "crucial event" passed to the server might not contain the `name` or `namespace` field.

File: test-rabbitmq-operator/build/Dockerfile [hunk 11]
```diff
@@ -26,6 +26,8 @@ FROM alpine:latest as etc-builder
 RUN echo "rabbitmq-cluster-operator:x:1000:" > /etc/group && \
     echo "rabbitmq-cluster-operator:x:1000:1000::/home/rabbitmq-cluster-operator:/usr/sbin/nologin" > /etc/passwd
 
+RUN sed -i 's/dl-cdn.alpinelinux.org/mirrors.tuna.tsinghua.edu.cn/g' /etc/apk/repositories
```
Comment (marshtompsxd) [inline test-rabbitmq-operator/build/Dockerfile:29]: I guess you do not need this one since you are also located in US now?

File: test-casskop-operator/test/atomic-1.yaml [hunk 12]
```diff
@@ -0,0 +1,25 @@
+project: casskop-operator
```
Comment (marshtompsxd) [inline test-casskop-operator/test/atomic-1.yaml:1]: Could you replace the atomic config with the obs gap config that can detect this bug since we plan to use obs gap to detect this bug?

File: test-casskop-operator/test/atomic-2.yaml [hunk 13]
```diff
@@ -0,0 +1,21 @@
+project: casskop-operator
```
Comment (marshtompsxd) [inline test-casskop-operator/test/atomic-2.yaml:1]: same above

File: csi-driver/install.sh (hunk 6 above)
Comment (marshtompsxd) [reply csi-driver/install.sh:1]: Just curious

Comment (laphets) [reply csi-driver/install.sh:1]: All the files in deploy is grabbed from https://github.com/kubernetes-csi/csi-driver-host-path/tree/master/deploy

File: test-casskop-operator/test/atomic-1.yaml (hunk 12 above)
Comment (laphets) [reply test-casskop-operator/test/atomic-1.yaml:1]: Let me do this in a separate commit

File: oracle.py [hunk 14]
```diff
@@ -208,5 +297,64 @@ def check(learned_side_effect, learned_status, testing_side_effect, testing_stat
             bug_report += "[DEBUGGING SUGGESTION] %s\n" % generate_debug_suggestion(
                 testing_config)
         bug_report = "\n[BUG REPORT]\n" + bug_report
+
+        if learned_resources != None:
```
Comment (marshtompsxd) [inline oracle.py:301]: @laphets 
One last question before merge: it seems that we still use status.json rather than resource.json to decide whether there is a bug or not, is that correct?

Comment (laphets) [reply oracle.py:301]: If there is `resource.json` generated from learning, the new oracle will take into effect.

---

For now, I only generate `resource.json` for certain workloads. In the future, after the new oracle feature is more stable, we can run a batch and generate `resource.json` for all our current workloads.

Comment (marshtompsxd) [reply oracle.py:301]: Sounds good. Let's merge this giant now.
